Key Components to be Fulfilled:
1. Loading and Preprocessing (2 marks):
● Load the California Housing dataset using the fetch_california_housing function
from sklearn.
● Convert the dataset into a pandas DataFrame for easier handling. Handle
missing values (if any) and perform necessary feature scaling (e.g.,
standardization).
● Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.
2. Regression Algorithm Implementation (5 marks):
● Implement the following regression algorithms:
○ Linear Regression
○ Decision Tree Regressor
○ Random Forest Regressor
○ Gradient Boosting Regressor
○ Support Vector Regressor (SVR)
● For each algorithm: Provide a brief explanation of how it works. Explain why it
might be suitable for this dataset.
3. Model Evaluation and Comparison (2 marks):
● Evaluate the performance of each algorithm using the following metrics:
○ Mean Squared Error (MSE)
○ Mean Absolute Error (MAE)
○ R-squared Score (R²)
● Compare the results of all models and identify: The best-performing algorithm
with justification and the worst-performing algorithm with reasoning.
 4. Timely Submission (1 mark):
● Submit your code in a Jupyter Notebook format via a GitHub link.
● Ensure that your code is well-documented and explanations are clear and
concise.
● Submission Guidelines: Use Python and Jupyter Notebook for implementation.
Submit the GitHub repository link containing the code and all necessary files.
Ensure proper documentation, including code comments and markdown
explanations.

In [ ]:
1. Loading and Preprocessing (2 marks):
● Load the California Housing dataset using the fetch_california_housing function
from sklearn.
● Convert the dataset into a pandas DataFrame for easier handling. Handle
missing values (if any) and perform necessary feature scaling (e.g.,
standardization).
● Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.

In [1]:
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
print("Dataset loaded successfully")

Dataset loaded successfully


In [3]:
import pandas as pd
df = pd.DataFrame(
    housing.data,
    columns=housing.feature_names
)
df["MedHouseVal"] = housing.target
df.head()
 

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


### **Explanation**
The California Housing dataset is a regression dataset available in the sklearn library. It contains different features related to houses in California and the corresponding median house value.
 

In [3]:
# Check for missing values
df.isnull().sum()

MedInc         0
HouseAge       0
AveRooms       0
AveBedrms      0
Population     0
AveOccup       0
Latitude       0
Longitude      0
MedHouseVal    0
dtype: int64

### **Explanation**
The dataset was checked for missing values using isnull().sum(). If missing values are present, they can be handled using appropriate techniques such as replacing them with the mean or median.
For this dataset, the missing-value check shows that there are no missing values, so no missing-value treatment is required.

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
# Separate features and target
X = df.drop("MedHouseVal", axis=1)
y = df["MedHouseVal"]
# Split dataset
X_train, X_test, y_train, y_test = train_test_split( X,y,  test_size=0.2, random_state=42)
# Standardization
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

### **Explanation**
Standardization converts the features to a common scale with approximately mean = 0 and standard deviation = 1.
It is especially important for algorithms such as SVR, because these algorithms are sensitive to differences in feature scales.
The assignment specifically requires handling missing values and performing necessary feature scaling.

2. Regression Algorithm Implementation (5 marks):
● Implement the following regression algorithms:
○ Linear Regression
○ Decision Tree Regressor
○ Random Forest Regressor
○ Gradient Boosting Regressor
○ Support Vector Regressor (SVR)
● For each algorithm: Provide a brief explanation of how it works. Explain why it
might be suitable for this dataset.

In [5]:
from sklearn.linear_model import LinearRegression
# Create model
linear_model = LinearRegression()
# Train model
linear_model.fit(X_train_scaled, y_train)
# Make predictions
y_pred_linear = linear_model.predict(X_test_scaled) 
print("Linear Regression completed")

Linear Regression completed


### **Explanation**
Linear Regression is a supervised learning algorithm that assumes a linear relationship between the input features and the target variable.
It tries to find the best-fitting line/equation that minimizes the difference between actual and predicted values.
Linear Regression is suitable as a baseline model for the California Housing dataset because it can identify linear relationships between the housing features and the median house value.
 

In [7]:
from sklearn.tree import DecisionTreeRegressor
# Create model
dt_model = DecisionTreeRegressor(random_state=42)
# Train model
dt_model.fit(X_train, y_train)
# Make predictions
y_pred_dt = dt_model.predict(X_test)
print("Decision Tree Regressor completed")

Decision Tree Regressor completed


### **Explanation**
A Decision Tree Regressor uses a tree-like structure to make predictions. It divides the data into smaller groups using different conditions and predicts the target value at the leaf nodes.
Decision Tree Regressor is suitable because it can capture non-linear relationships between housing features and house prices without requiring feature scaling.
 

In [9]:
from sklearn.ensemble import RandomForestRegressor
# Create model
rf_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)
# Train model
rf_model.fit(X_train, y_train)
# Make predictions
y_pred_rf = rf_model.predict(X_test) 
print("Random Forest Regressor completed")
 

Random Forest Regressor completed


### **Explanation**
Random Forest Regressor is an ensemble learning algorithm that combines predictions from multiple decision trees.Each tree makes a prediction, and the predictions are combined to produce the final result.
Random Forest Regressor is suitable because it combines multiple decision trees and can capture complex non-linear relationships in the housing data while reducing overfitting compared with a single decision tree.

In [10]:
from sklearn.ensemble import GradientBoostingRegressor
# Create model
gb_model = GradientBoostingRegressor(
    random_state=42
)
# Train model
gb_model.fit(X_train, y_train)
# Make predictions
y_pred_gb = gb_model.predict(X_test)
print("Gradient Boosting Regressor completed")

Gradient Boosting Regressor completed


### **Explanation**
Gradient Boosting Regressor builds multiple decision trees sequentially. Each new tree attempts to correct the errors made by the previous trees.
Gradient Boosting Regressor is suitable because it builds trees sequentially to correct previous errors and can model complex relationships between housing features and median house values.
 

In [11]:
from sklearn.svm import SVR
# Create model
svr_model = SVR()
# Train model using scaled data
svr_model.fit(X_train_scaled, y_train)
# Make predictions
y_pred_svr = svr_model.predict(X_test_scaled)
print("SVR completed")

SVR completed


### **Explanation**
Support Vector Regressor (SVR) is a supervised learning algorithm that tries to find a function that predicts the target values while keeping prediction errors within a specified margin.
SVR is suitable because it can model complex non-linear relationships between the input features and house values. Standardization also makes the features appropriate for SVR because SVR is sensitive to feature scales.
 

3. Model Evaluation and Comparison (2 marks):
● Evaluate the performance of each algorithm using the following metrics:
○ Mean Squared Error (MSE)
○ Mean Absolute Error (MAE)
○ R-squared Score (R²)
● Compare the results of all models and identify: The best-performing algorithm
with justification and the worst-performing algorithm with reasoning.

In [12]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score
# Store predictions
predictions = {
    "Linear Regression": y_pred_linear,
    "Decision Tree": y_pred_dt,
    "Random Forest": y_pred_rf,
    "Gradient Boosting": y_pred_gb,
    "SVR": y_pred_svr
}
# Evaluate each model
results = []
for model_name, y_pred in predictions.items():
 
    mse = mean_squared_error(y_test, y_pred)
    mae = mean_absolute_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
 
    results.append({
        "Model": model_name,
        "MSE": mse,
        "MAE": mae,
        "R2 Score": r2
    })
# Create comparison DataFrame
results_df = pd.DataFrame(results)
results_df

,Model,MSE,MAE,R2 Score
0,Linear Regression,0.555892,0.533200,0.575788
1,Decision Tree,0.495235,0.454679,0.622076
2,Random Forest,0.255368,0.327543,0.805123
3,Gradient Boosting,0.293997,0.371643,0.775645
4,SVR,0.357004,0.398599,0.727563


In [16]:
# Round the results for easier reading
results_df.round(4)

,Model,MSE,MAE,R2 Score
0,Linear Regression,0.5559,0.5332,0.5758
1,Decision Tree,0.4952,0.4547,0.6221
2,Random Forest,0.2554,0.3275,0.8051
3,Gradient Boosting,0.2940,0.3716,0.7756
4,SVR,0.3570,0.3986,0.7276


In [17]:
best_model = results_df.loc[
    results_df["R2 Score"].idxmax()
]
print("Best Performing Model:")
print(best_model)
 

Best Performing Model:
Model       Random Forest
MSE              0.255368
MAE              0.327543
R2 Score         0.805123
Name: 2, dtype: object


In [19]:
worst_model = results_df.loc[
    results_df["R2 Score"].idxmin()
]
print("Worst Performing Model:")
print(worst_model)

Worst Performing Model:
Model       Linear Regression
MSE                  0.555892
MAE                    0.5332
R2 Score             0.575788
Name: 0, dtype: object


### **Explanation**
The `mean_squared_error`, `mean_absolute_error`, and `r2_score` functions are imported from `sklearn.metrics` to evaluate the performance of the regression models. The predictions of all five regression models are stored in a dictionary called `predictions`. A `for` loop is then used to calculate the MSE, MAE, and R² score for each model.The calculated evaluation metrics are stored in a list called `results`. Finally, the list is converted into a Pandas DataFrame named `results_df`, which makes it easier to compare the performance of all five regression models.
A lower MSE and MAE indicate better performance, while a higher R² score indicates better performance.

### **Comparison and Conclusion**
 
Based on the evaluation results, Random Forest Regressor is the best-performing model. It has the lowest MSE (0.2554), the lowest MAE (0.3275), and the highest R² score (0.8051). Therefore, it gives the best overall performance among the five models.
 
Linear Regression is the worst-performing model because it has the highest MSE (0.5559), the highest MAE (0.5332), and the lowest R² score (0.5758). Therefore, it gives the weakest overall performance among the five models.